### ISSUE TITLE
Refactoring of MapStore adapter models architecture

### ISSUE THREAD

**Issue by giohappy** [geosolutions-it/nexus-geonode#106]
The MapStore client needs to store some specific information, particularly it's json "blob", which goes beyond what the GeoNode [Map](https://github.com/GeoNode/geonode/blob/master/geonode/maps/models.py#L55) model can hold.

These data have been implemented as client models inside the mapstore adapter [models.py](https://github.com/GeoNode/geonode-mapstore-client/blob/master/mapstore2_adapter/api/models.py) module, to avoid touching the GeoNode Map model and to support the initial idea of being able to use the mapstore adapter even without GeoNode. The latter also explains why we have two modules, the geonode_mapstore_client en the mapstore2_adapter (which initially were managed in separate repos).
The latter idea never evolved, and the experience with the geoapps module shows that using a generic data holder inside the core models give, obviously, significant benefits:

- let the client use a single API (the GeoNode core one) instead of having to request both the core API to retrieve GeoNode Map informtion and the masptore adapter API to obtain the specific mapstore information
- remove the need of having to sync the MapStoreResource and Map models on creation, update, deletion (acrually I spotted a bug: a MapStore Resource is not deleted now when a Map is deleted!)

The proposal here is to get rid of the mapstore models and move the [MapStoreResource.data](https://github.com/GeoNode/geonode-mapstore-client/blob/master/mapstore2_adapter/api/models.py#L48) (-> MapStoreData class) field to a Map.blob field, similar to what we have with [Geoapps.data](https://github.com/GeoNode/geonode/blob/master/geonode/geoapps/models.py#L76) (->GeoAppData).

This of course inpacts many places:

 - client requests must be remapped and adapted to use the core API only instead of the mapstore adapter API
- mapstore2 adapter code depending on the current models must be adapter

To avoid breaking the client we might initially adapt the MapStoreResourceViewSet to delegate to the refactored models. When the client will be adapter we can then remove the mapstore adapter API.

**NOTICE**: GeoNode Map and MapLayer derive from GPX* classes, which refer to the previous client architecture (which used, by default, the GeoExplorer framework). As a further step we could probably refactor these concepts too, but I think we can and should make one step at a time to reduce the complexity and the risks connected to these changes.

**Comment by afabiani** [comment]
All that class infrastructure inherited from `utils.py` (`GXP*Map`, `GXP*Layer`, ...) is now completely unesful and could be easily removed by adding the missing attributes back to the base classes.

Pay attention to two things mainly:

1. There are several utility methods handling some sort of predefined JSON configurations, both for the maps and the layers. Those are used in different parts of the code; as an instance to create an empty map, to create the context for a layer viewer, etc... and also the same assumption are done into the `adapter serializer & plugin`. IMHO we don't need such meta-model which was required before to wrap the GeoExt objects into Python ones and viceversa. So we could get rid potentially of a lot of unuseful and unused classes.

2. Be careful on removing such classes; most of the informations stored on their attributes are used to persist the original/current real map configurations. As an instance the legends of the layers, the base urls, the types (background layers or something else), and so on. In some cases there are passages where the code creates a sort of `layer capabilities` storing specific information, like the time instants, if the layer is a mosaic or not, the feature-type-info templates, etc... Again, like in (1) the `adapter serializer & plugin` often relies on those information.

Long story short, we can make everything much more light and remove a lot of passages which are not necessary anymore, but we need to pay a lot of attention to not lose crucial information along the way.

Most probably the entire `adapter serializer & plugin` logic must be revised too. That one was supposed to convert back and forth such old-fashion json style configurations into python objects, and python objects into MapStore configurations.

**Comment by mattiagiupponi** [comment]
So briefly the idea is to:
- `Map` model in GeoNode: create a new field as a class with blob and resource (liked for Geoapps )
-- `MapStoreResource` -> `Map` & `MapStoreData` -> `MapData` (new class)
- Move/rewrite the API from `MapstoreAdapter` into `GeoNode`
- Migrate (if possible) the blob available that follows the old logic in the new one in order to dismiss totally the old API (or maintains both API?)
- Remove the inherit of `GXP*` by adding the fields to the original model (`Map`)
- Don't lose information for `adapter serializer & plugin`
- Massive tests

@giohappy is this briefly the issue?

**Comment by giohappy** [comment]
As agreed @mattiagiupponi for the moment it's enough to have the old MapStore2 API moved to the the core API v2.
For GeoNode 3.x we must also ensure that __GEONODE__CONFIG__ still works, i.e. the template context is created correctly.

The deep refactoring of the Map models (including the GXP stuff) and the Mapstore adapter can be postponed.

**Comment by mattiagiupponi** [comment]
Hi @giohappy 
Most of the refactoring has been completed and now the maps are working through the new endpoint in v2.
I also delete everything that is related to `MapstoreResource` included:
- Model
- Viewsets
- Serializers
- URLs

I made some tests on the local environment and looks like that works fine (from a user POV).
Now the following tasks are missing:
- Fix tests in GeoNode (since the MapSerializer is different)
- Fix tests in client
- QA

I'm gonna create some issues attached in this epic for a better estimation

**Comment by giohappy** [comment]
> I think that will be required a migration file in order to move the configuration from the map store tables to the new ones

@mattiagiupponi  definietely!

**Comment by mattiagiupponi** [comment]
Mapstore refactor PR (still a WIP) -> https://github.com/GeoNode/geonode-mapstore-client/pull/171
Geonode refactor PR (still a WIP) -> https://github.com/GeoNode/geonode/pull/7531

**Comment by mattiagiupponi** [comment]
PRs removed from WIP
